### PR TITLE
[DOCS] #5310 - Overhaul borg help patterns

### LINKED ISSUE #5310: How to back up only certain files (e.g. *.jpg)
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Question

#### Your borg version (borg -V).

borg 1.1.13

#### Operating system (distribution) and version.

Ubuntu 18.04.4

#### Hardware / network configuration, and filesystems used.

Intel(R) Core(TM)2 Duo CPU E8400 @ 3.00GHz
16 GB RAM
ZFS 5 x 3-way mirrors, 22 TiB usable

#### How much data is handled by borg?

15 TiB

## Question

Is this possible? I can't find a reference to a command-line flag that will do this, or related FAQ, doc, or issue.

The reason I'm doing this, is to back up my most important files first, which I can get most optimally "close enough" with wildcard matching.

I understand I can generate a separate "includes" file but if there's a way to keep it native, that would be better for my use case.

Many thanks.

Update: I realize there actually is no `--files-from` flag or similar functionality (and `--patterns-from` is marked experiments - and seems a little dicey to include millions of files).

### PR DESCRIPTION
**Pull request by fantasya-pbem**
My first shot to improve the patterns help a bit. I tried to keep it short.
- Start with explaination where patterns can be used.
- Emphasize that leading separator is trimmed.
- List patterns file prefixes one by one to make it more readable.
- Add the tip for testing patterns.
- Keep examples short.

I haven't changed existing examples a lot, but I tested all aspects with different pattern styles.

fixes #5310

### CODE REVIEW

**Comment by fantasya-pbem** [comment]
Of course I checked "build_usage", "build_man" and "make html"...

File: src/borg/archiver.py [hunk 1]
```diff
@@ -2626,34 +2626,56 @@ def do_break_lock(self, args, repository):
...
+        A more general and easier to use way to define filename matching patterns
+        exists with the ``--pattern`` and ``--patterns-from`` options. Using
+        these, you may specify the backup roots, default pattern styles and
+        patterns for inclusion/exclusion.
 
-        .. note::
+        Root path prefix ``R``
+            A root path starts with the prefix `R`, followed by a path (a plain
+            path, not a file pattern). Use this prefix to move root paths from
+            command line arguments to a patterns file.
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2637]: maybe it should use "recursion root paths" at least in the explanation.

File: src/borg/archiver.py [hunk 2]
```diff
@@ -2626,34 +2626,56 @@ def do_break_lock(self, args, repository):
...
+        Exclude no-recurse pattern prefix ``!``
+            Use the prefix `!`, followed by a pattern, to define an exclusion
+            that does not recurse into subdirectories.
+
+        Inclusion patterns are useful to include paths that are contained in an
+        excluded path. The first matching pattern is used. so if an include
+        pattern matches before an exclude pattern, the file is backed up. If an
+        exclude-norecurse pattern matches a directory, it won't recurse into it
+        (which saves time), but won't discover any potential matches for include
+        rules below that directory.
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2664]: change order:

1. exclude
2. exclude no-recurse
3. include

and then explain why include is useful (maybe even within the include explanation).

File: src/borg/archiver.py [hunk 3]
```diff
@@ -2626,34 +2626,56 @@ def do_break_lock(self, args, repository):
...
-        patterns. The pattern style can be set via the `P` prefix.
+        This will print the considered files one per line, prefixed with a
+        character that indicates the action (e.g. 'x' for excluding, see
+        **Item flags** in `borg create` usage docs).
+
+        .. note::
+
+            It's possible that a sub-directory/file is matched while parent
+            directories are not. In that case, parent directories are not backed
+            up thus their user, group, permission, etc. can not be restored.
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2678]: ... backed up **and** thus ... ?

File: src/borg/archiver.py [hunk 4]
```diff
@@ -2665,42 +2687,40 @@ def do_break_lock(self, args, repository):
             # note: using = is essential to avoid cmdline argument parsing issues.
             borg create --pattern=+pics/2018/good --pattern=-pics/2018 repo::arch pics
 
+            # backup only JPG/JPEG files (case insensitive) in all home directories:
+            borg create --pattern 're:\\.jpe?g(?i)' repo::arch /home
+
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2692]: maybe a more realistic example would be to exclude all *.ISO (they usually are downloads **and** big).

guess most people do not **only** want to backup their jpegs, but also other files in their homes.

File: src/borg/archiver.py [hunk 5]
```diff
@@ -2665,42 +2687,40 @@ def do_break_lock(self, args, repository):
...
             # use a file with patterns:
-            borg create --patterns-from patterns.lst repo::arch
+            borg create --patterns-from patterns.lst repo::arch /
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2694]: is the / required? the recursion root can also be given in the patterns.lst.

File: src/borg/archiver.py [hunk 6]
```diff
@@ -2665,42 +2687,40 @@ def do_break_lock(self, args, repository):
...
 
         The patterns.lst file could look like that::
 
-            # "sh:" pattern style is the default, so the following line is not needed:
+            # sh: pattern style is the default, so the following line is not needed:
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2698]: is that an improvement? i guess the quotes were there intentionally because the colon is not punctuation in that sentence.

File: src/borg/archiver.py [hunk 7]
```diff
@@ -2665,42 +2687,40 @@ def do_break_lock(self, args, repository):
...
             # include susans home
             + home/susan
             # also back up this exact file
             + pf:home/bobby/specialfile.txt
             # don't backup the other home directories
             - home/*
-            # don't even look in /proc
-            ! proc
+            # don't even look in /dev, /proc, /run, /sys, /tmp
+            ! re:^(dev|proc|run|sys|tmp)/
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2709]: if there is a slash at the end, borg might try to look "into" these directories, because the directory was not matched by the expression (yet), I guess?

File: src/borg/archiver.py [hunk 8]
```diff
@@ -2626,34 +2626,56 @@ def do_break_lock(self, args, repository):
...
 
-            It's possible that a sub-directory/file is matched while parent directories are not.
-            In that case, parent directories are not backed up thus their user, group, permission,
-            etc. can not be restored.
+            $ borg create --dry-run --list --patterns-from patterns.txt /repo::arch
 
-        Note that the default pattern style for ``--pattern`` and ``--patterns-from`` is
-        shell style (`sh:`), so those patterns behave similar to rsync include/exclude
-        patterns. The pattern style can be set via the `P` prefix.
+        This will print the considered files one per line, prefixed with a
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2670]: s/print/list/

**Review by ThomasWaldmann — COMMENTED** [review]
thanks for the PR, some feedback.

File: src/borg/archiver.py [hunk 9]
```diff
@@ -2485,35 +2485,36 @@ def do_break_lock(self, args, repository):
...
-        Starting with Borg 1.2, discovered fs paths are normalised, have leading
-        slashes removed and then are matched against your patterns.
-        Note: You need to review your include / exclude patterns and make
-        sure they do not expect leading slashes. Borg can only deal with this
-        for some very simple patterns by removing leading slashes there also.
-
-        If followed by a colon (':') the first two characters of a pattern are
+        When specifying one or more file paths in a Borg command that supports
+        patterns for the respective option or argument, you can apply the
+        patterns described here to filter out desired files. Patterns can be
```
[the remark this answers is not in the document]
Comment (fantasya-pbem) [reply src/borg/archiver.py:2490]: I'll rephrase this to "...patterns described here to include only desired files and/or exclude unwanted ones."

File: src/borg/archiver.py [hunk 10]
```diff
@@ -2485,35 +2485,36 @@ def do_break_lock(self, args, repository):
...
+
+        - for ``--exclude`` option,
+        - in the file given with ``--exclude-from`` option,
+        - for ``--pattern`` option,
+        - in the file given with ``--patterns-from`` option and
+        - for ``PATH`` arguments that explicitly support them.
+
+        Borg always stores all file paths normalized and relative to the
+        current recursion root. The recursion root is also named ``PATH`` in
+        Borg commands that do a file search like `borg create`, so don't
```
[the remark this answers is not in the document]
Comment (fantasya-pbem) [reply src/borg/archiver.py:2501]: Discovery sounds good.

There has always been a mix of double/single backticks. I'm not sure if we should use only one or the other.

File: src/borg/archiver.py [hunk 11]
```diff
@@ -2485,35 +2485,36 @@ def do_break_lock(self, args, repository):
...
+        Starting with Borg 1.2, paths that are matched against patterns always
+        appear relative. If you give ``/absolute/`` as root, the paths going
+        into the matcher will look relative like `absolute/.../file.ext`. If you
+        give ``../some/path`` as root, the paths will look like
+        `some/path/.../file.ext`.
+
+        File patterns support these styles: **fnmatch**, **shell**, **regular**
+        **expressions**, **path prefixes** and **path full-matches**. By default,
+        fnmatch is used for ``--exclude`` patterns and shell style is used for
+        the ``--pattern`` option. For commands that support patterns in their
```
[the remark this answers is not in the document]
Comment (fantasya-pbem) [reply src/borg/archiver.py:2513]: Yes, this can be shortened a bit.

File: src/borg/archiver.py [hunk 12]
```diff
@@ -2540,16 +2541,15 @@ def do_break_lock(self, args, repository):
             is that the pattern may include `**/` for matching zero or more directory
             levels, `*` for matching zero or more arbitrary characters with the
             exception of any path separator. A leading path separator is always removed.
+            Shell-style patterns behave similar to rsync include/exclude patterns.
```
[the remark this answers is not in the document]
Comment (fantasya-pbem) [reply src/borg/archiver.py:2544]: I agree, we should not reference to rsync. It does not clarify anything here.

File: src/borg/archiver.py [hunk 13]
```diff
@@ -2540,16 +2541,15 @@ def do_break_lock(self, args, repository):
...
 
-        Regular expressions, selector `re:`
+        `Regular expressions <https://docs.python.org/3/library/re.html>`_, selector `re:`
             Regular expressions similar to those found in Perl are supported. Unlike
```
[the remark this answers is not in the document]
Comment (fantasya-pbem) [reply src/borg/archiver.py:2547]: ok

File: src/borg/archiver.py (hunk 1 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2637]: I didn't find the right words. Nice!

File: src/borg/archiver.py (hunk 2 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2664]: ok

File: src/borg/archiver.py (hunk 8 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2670]: ok

File: src/borg/archiver.py (hunk 3 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2678]: ok

File: src/borg/archiver.py (hunk 4 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2692]: I just realize that there is a missing '+' in 're:'...
I wanted to include a simple example for the use case of #5310 question (back up files of distinct extension).
Of course, we can add more common useful examples. I'll add something...

File: src/borg/archiver.py (hunk 5 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2694]: It's not required, but showcases a patterns file without R prefix. The other example below features roots in the file only.

File: src/borg/archiver.py (hunk 6 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2698]: ok, reverting

File: src/borg/archiver.py (hunk 7 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2709]: That's right, but without / it would exclude also files like "/device". Not very likely, but...
I remove the / and add a comment with a warning.

**Comment by fantasya-pbem** [comment]
Please wait until I re-checked the built HTML and verified examples again.

File: src/borg/archiver.py (hunk 5 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py:2689]: I added a comment to clarify this.

File: src/borg/archiver.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (fantasya-pbem) [reply src/borg/archiver.py]: I now use double backticks everywhere except for borg commands, these appear underlined and look like links then.

**Comment by fantasya-pbem** [comment]
Ok, please continue your review.

File: src/borg/archiver.py [hunk 14]
```diff
@@ -2624,36 +2617,58 @@ def do_break_lock(self, args, repository):
...
 
-        Inclusion patterns are useful to include paths that are contained in an excluded
-        path. The first matching pattern is used so if an include pattern matches before
-        an exclude pattern, the file is backed up. If an exclude-norecurse pattern matches
-        a directory, it won't recurse into it and won't discover any potential matches for
-        include rules below that directory.
+        Exclude pattern prefix ``-``
+            Use the prefix ``-``, followed by a pattern, to define an exclusion
+            for ``--patterns`` / ``--patterns-from`` and ``--exclude`` /
+            ``--exclude-from``.
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2641]: i don't think we support prefixing a minus sign for `--exclude*`.

File: src/borg/archiver.py [hunk 15]
```diff
@@ -2624,36 +2617,58 @@ def do_break_lock(self, args, repository):
...
 
-        .. note::
+        Exclude no-recurse pattern prefix ``!``
+            Use the prefix ``!``, followed by a pattern, to define an exclusion
+            that does not recurse into subdirectories. This saves time, but
+            prevents include patterns to match any files in subdirectories.
+
+        Include pattern prefix ``+``
+            Use the prefix ``+``, followed by a pattern, to define include files.
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py:2649]: "include files" have a special meaning for software developers.

so i'ld rather say "inclusions" or "files to include".

File: src/borg/archiver.py (hunk 14 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py]: Yes, that makes no sense.

File: src/borg/archiver.py (hunk 15 above)
Comment (fantasya-pbem) [reply src/borg/archiver.py]: Rephrased!

**Review by ThomasWaldmann — APPROVED** [review]

**Comment by ThomasWaldmann** [comment]
thanks!
